### DISCUSSION TITLE
Moose commit history

### DISCUSSION THREAD

**Discussion by singhgp4321** [idaholab/moose#16974] [Q&A General] [answered]
Hi,

I noticed that the commit history of moose submodule in Bison app is different for different versions of moose. Of course the latest version of moose will have additional commits, but for the common time range some of the commits found in one version of moose are not present in the other version of moose. 

I am wondering if someone could explain why the commit history for moose differ from version to version. I can provide additional information if needed; let me know.

Thanks,
Gyan

**Comment by loganharbour** [comment, marked as answer]
We utilize the method of continuous integration when developing MOOSE and the MOOSE-based applications (see [this](https://en.wikipedia.org/wiki/Continuous_integration) wikipedia article for more info). We update the software many times a day, and when testing passes for a change in MOOSE, we update many of the MOOSE-based applications on the same day once their tests pass with the changes to MOOSE. This allows us to get updates to applications as soon as possible, whether it be bug fixes or new features.

The master branch of BISON has its MOOSE submodule updated once all of its own tests pass with the new changes to MOOSE. With this, you can grab any commit of BISON, and have a version of MOOSE (noted within the submodule) that is guaranteed to be compatible. 

This is why you will see the MOOSE submodule within BISON (and many other MOOSE-based applications) change quite a bit: we update BISON as soon as reasonably possible with newer versions (more recent commits) of MOOSE.

Bottom line: whenever you update an application, be sure to update the MOOSE submodule if it has one!

**Reply by gambka** [reply]
@loganharbour the issue @singhgp4321 is having isn't related to updating the submodule due to the continuous integration. It is due to commits that were in an older submodule of MOOSE are NOT found (the sha's are not shown) in the git log of the submodule in a NEWER version. Say we checkout a version of the BISON with a submodule of moose an update to the version of moose packaged with that version of the downstream app. Now lets say we take a version of BISON 3 days later and update to the moose submodule packaged with that version. If you go into the moose directory and do a git log in the newer version I would suspect all MOOSE commits to be the same as the older version plus any new commits in those 3 days. This is not the case. The old submodule will have MOOSE commits on say Jan 4. But then if you update to the new moose submodule in the newer version and the submodule git log shows zero commits on Jan 4. Why would the Jan 4 commit inside the MOOSE submodule be present in the older version of BISON but not the newest? The submodule should have been updated and new commits added, but old commits should not be removed.

**Reply by GiudGiud** [reply]
git log doesn't return a linear in time history, are you sure the Jan 4 commits aren't just further down the log?

Either way if you have a specific example of a commit that is not longer there after a submodule update then we can look into it.

**Reply by gambka** [reply]
I'm pretty sure the sha was not found in one but found in the other. I'll have @singhgp4321  provide his example. I know the dates do not have to be linear if someone has been amending a commit over time and it gets merged. And Jan 4 was just an example but the same idea applies.

**Reply by loganharbour** [reply]
We never rewrite history.

The only time I could see this is an issue is with a manual submodule update with a commit that comes from a PR instead of within next/devel/master. Github is smart enough to let you use a sha from a PR when pulling from a repo. That sha could no longer exist.

It shouldn't be hard to write a quick python script to parse the log and find these. I don't know if this is the issue, but this is the only case I could think of. If this is indeed the problem, we should cause prechecks to fail if the hash isn't associated with moose/next.

**Reply by GiudGiud** [reply]
That'd be great. 
Just a side note that the dates are generally not linear, not just if commits have been amended. Rebasing for example will keep the original dates of your commits, yet place them all at the end of the log.

**Reply by singhgp4321** [reply]
Here are the commit histories for two versions of Moose for two corresponding versions of Bison. You'll notice that Jan 4 commit shows up in one version's commit history but not in the other one: [moose_commit_histories.txt](https://github.com/idaholab/moose/files/5991817/moose_commit_histories.txt)

**Reply by GiudGiud** [reply]
If you checkout 59f7e5b61b9bba11ebab9f506f356693e973b12f and look at the log you will find all the commits that are in your text file. You might have to go a bit far in the log, I recommend using `git log --pretty=format:'%h : %s' > log.log`

From git log at the first commit in your file, the January 4 commits in bold.

59f7e5b61b : Merge commit '44404267f136d11f3a09c8220ea98c6d4f809034'
44404267f1 : Merge pull request #16618 from lindsayad/check-leaks
96a63b894d : Merge pull request #16404 from crswong888/listing-improvement-16401
2e197b23d5 : Merge pull request #16588 from aeslaughter/reporter-declare-initialSetup-16586
3164e97de1 : Update libmesh
bb9d19be9f : Only populate attributes if supplied and reincorporate the attributes override into listing extension
310686b1af : Updated documentation for listing extension
9fc3b358d4 : created unit tests for listing extension
ba6897ed57 : change names of 'prefix' and 'suffix' settings to 'prepend' and 'append' in extractContent() to avoid conflicts
56dd034cbc : Fix bug with float numbers - reset counter on every tokenization so that it doesn't accumulate during a live serve, which leads to higher and higher numbers.
aea2874d72 : Execute pre- and post-read methods for MooseDocsTestCase
3b70222a34 : Removing listing gold
ba54103ce1 : more refactoring of listing extension
9f4a6a05bb : Re-factor ability to remove blocks and params from hit files
e163c9f75d : listing extension can now exclude blocks and parameters that otherwise would be rendered
e572e43794 : Merge pull request #16157 from lindsayad/isnt-prepared-mesh-api
**eec233d851 : Merge commit '3877bd279e97bab2a579b926ecd33a543ea2fa2a'**
211f7da580 : Initialize output lists for each call to output
e010c683fb : Create test to demonstrate issue #16586
44a89d7cb9 : Merge pull request #16570 from loganharbour/modules_unit
46327bf62c : Use MeshBase data to inform us when we should prepare
198485f2a7 : Add test
f9af866daa : Merge pull request #16397 from lindsayad/auto-full-matrix-16396
**3877bd279e : Merge pull request #16463 from cbolisetti/csvsampler**
**8444a70b28 : Merge pull request #16527 from milljm/16513-moose-pyhit-tlc**
**071d52903b : Merge pull request #16475 from arovinelli/nl_forced_its**
e5cb79c9e9 : Merge commit 'bbd2e27f25001c4834201d1d3ff5bb4fae46c84e'
bbd2e27f25 : Merge pull request #16597 from loganharbour/pandas_update
b319bbde19 : Resolve comparison warnings for stochastic tools unit tests
6b0743b89f : Resolve comparison warnings for porous flow unit tests
da567e7525 : Resolve comparison warnings in geochemistry unit tests

**Reply by singhgp4321** [reply]
Thanks Guillaume for the clarification! Yes, I do see the Jan 4 commits after going far down in the log; for some reason I could not view the log down enough earlier. Also, I am realizing that `git log` isn't the best way to look at the log. The `pretty` and `graph` arguments provide a much cleaner and clearer history of the commits.  `git log --since "JAN 01 2021" --until "JAN 14 2021" --pretty=format:"%h %cd %s" --graph` provides what I was looking for. Thanks!
